src/search/pdbs/pattern_collection_generator_multiple.cc

```diff
@@ -0,0 +1,310 @@
+    const utils::CountdownTimer &timer) {
+    const Pattern &pattern = pattern_info.get_pattern();
+    if (verbosity >= utils::Verbosity::DEBUG) {
+        utils::g_log << "generated pattern " << pattern << endl;
+    }
+    if (generated_patterns.insert(move(pattern)).second) {
+        // RCG generated a new pattern. Reset stagnation_start_time.
+        stagnation_start_time = -1;
+        shared_ptr<PatternDatabase> pdb = pattern_info.get_pdb();
+        remaining_collection_size -= pdb->get_size();
+        generated_pdbs->push_back(move(pdb));
+    } else {
+        // Pattern is not new. Set stagnation start time if not already set.
+        if (stagnation_start_time == -1) {
+            stagnation_start_time = timer.get_elapsed_time();
+        }
+    }
+}
+
+bool PatternCollectionGeneratorMultiple::collection_size_limit_reached() const {
+    if (remaining_collection_size <= 0) {
+        /*
+          This value can become negative if the given size limits for
+          pdb or collection size are so low that compute_pattern already
+          violates the limit, possibly even with only using a single goal
+          variable.
+        */
+        if (verbosity >= utils::Verbosity::NORMAL) {
+            utils::g_log << "collection size limit reached" << endl;
+        }
+        return true;
+    }
+    return false;
+}
+
+bool PatternCollectionGeneratorMultiple::time_limit_reached(
+    const utils::CountdownTimer &timer) const {
+    if (timer.is_expired()) {
+        if (verbosity >= utils::Verbosity::NORMAL) {
+            utils::g_log << "time limit reached" << endl;
+        }
+        return true;
+    }
+    return false;
+}
+
+bool PatternCollectionGeneratorMultiple::check_for_stagnation(
+    const utils::CountdownTimer &timer) {
+    // Test if no new pattern was generated for longer than stagnation_limit.
+    if (stagnation_start_time != -1 &&
+        timer.get_elapsed_time() - stagnation_start_time > stagnation_limit) {
+        if (enable_blacklist_on_stagnation) {
+            if (blacklisting) {
+                if (verbosity >= utils::Verbosity::NORMAL) {
+                    utils::g_log << "stagnation limit reached "
+                                 << "despite blacklisting, terminating"
+                                 << endl;
+                }
+                return true;
+            } else {
+                if (verbosity >= utils::Verbosity::NORMAL) {
+                    utils::g_log << "stagnation limit reached, "
+                                 << "enabling blacklisting" << endl;
+                }
+                blacklisting = true;
+                stagnation_start_time = -1;
+            }
+        } else {
+            if (verbosity >= utils::Verbosity::NORMAL) {
+                utils::g_log << "stagnation limit reached, terminating" << endl;
+            }
+            return true;
+        }
+    }
+    return false;
+}
+
+static PatternCollectionInformation get_pattern_collection(
```

Comment:
This looks like a general utility function for PDBs. Is this useful elsewhere?

Reply:
I'm not aware of any place, but we could still move it to `utils`, if you like.

Reply:
Yes, I think it would be a better place for it.